components/butterfly/src/network.rs

```diff
@@ -80,3 +88,219 @@ pub trait Network: Send + Sync + Debug + 'static {
+            Error::GossipSenderSendError(e.description().to_owned())
+        })
+    }
+}
+
+impl GossipReceiver for GossipZmqSocket {
+    fn receive(&self) -> Result<Vec<u8>> {
+        self.zmq.recv_bytes(0).map_err(|e| {
+            Error::GossipReceiverReceiveError(e.description().to_owned())
+        })
+    }
+}
+
+/// An implementation of the `Network` trait that creates
+/// `SwimUdpSocket` instances for SWIM communication, and
+/// `GossipZmqSocket` instances for gossip communication.
+pub struct RealNetwork {
+    swim_addr: SocketAddr,
+    gossip_addr: SocketAddr,
+    push_socket_linger: i32,
+    zmq_context: ServerContext,
+}
+
+impl RealNetwork {
+    /// Create an instance of `RealNetwork` to be used by clients. It
+    /// sets linger time for push zmq gossip sockets to be
+    /// indefinite. Use this instance to get pull or push gossip
+    /// sockets.
+    ///
+    /// For getting SWIM channels or gossip receivers, get an instance
+    /// with `new_for_server`.
+    pub fn new_for_client() -> Self {
+        // The client only sends rumors through the gossip sender
+        // socket, so we pass some throw away address as arguments for
+        // SWIM socket and gossip receiver socket addresses.
+        Self::new(Self::throw_away_addr(), Self::throw_away_addr(), -1)
+    }
+
+    /// Create an instance of `RealNetwork` to be used by servers.
+    pub fn new_for_server(swim_addr: SocketAddr, gossip_addr: SocketAddr) -> Self {
+        Self::new(swim_addr, gossip_addr, 1000)
+    }
+
+    fn new(swim_addr: SocketAddr, gossip_addr: SocketAddr, push_socket_linger: i32) -> Self {
+        RealNetwork {
+            swim_addr: swim_addr,
+            gossip_addr: gossip_addr,
+            push_socket_linger: push_socket_linger,
+            zmq_context: ServerContext(UnsafeCell::new(zmq::Context::new())),
+        }
+    }
+
+    fn throw_away_addr() -> SocketAddr {
+        let ip = IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0));
+        SocketAddr::new(ip, 0)
+    }
+
+    fn context_mut(&self) -> &mut zmq::Context {
+        self.zmq_context.as_mut()
+    }
+}
+
+// Implementing Debug trait explicitly to avoid debug output for zmq_context
+impl Debug for RealNetwork {
+    fn fmt(&self, f: &mut Formatter) -> StdResult<(), FmtError> {
+        write!(
+            f,
+            "RealNetwork {{ swim_addr: {:?}, gossip_addr: {:?}, push_socket_linger: {:?}, \
+             zmq_context: <skipped> }}",
+            self.swim_addr,
+            self.gossip_addr,
+            self.push_socket_linger,
+        )
+    }
+}
+
+impl Network for RealNetwork {
+    type SwimChannel = SwimUdpSocket;
+    type GossipReceiver = GossipZmqSocket;
+    type GossipSender = GossipZmqSocket;
+
+    fn get_swim_addr(&self) -> SocketAddr {
+        self.swim_addr.clone()
+    }
+
+    fn get_swim_port(&self) -> u16 {
+        self.swim_addr.port()
+    }
+
+    fn get_swim_channel(&self) -> Result<SwimUdpSocket> {
```

Comment:
The `get` naming implies (to me, anyway) that the type is giving you access to some data it owns. These `get` functions, though, are more like constructors. Naming them accordingly could help clarify the code.